agent/src/agent/monitoring/streamsets.py

```diff
@@ -68,8 +71,42 @@ def _pull_kafka_metrics(streamsets_: streamsets.StreamSets):
 
 
 def pull_metrics():
-    for streamsets_ in streamsets.repository.get_all():
-        _pull_system_metrics(streamsets_)
-        _pull_kafka_metrics(streamsets_)
-    for pipeline_ in pipeline.repository.get_all():
-        _pull_pipeline_metrics(pipeline_)
+    streamsets_ = streamsets.repository.get_all()
+    _process_streamsets_metrics(
+        streamsets_=streamsets_,
+        asynchronous=len(streamsets_) >= 2
+    )
+    pipelines = pipeline.repository.get_all()
+    _process_pipeline_metrics(
+        pipelines=pipelines,
+        asynchronous=len(pipelines) >= 2
+    )
+
+
+def _process_pipeline_metrics(pipelines: List[pipeline.Pipeline], asynchronous: bool = False) -> None:
+    if not asynchronous:
+        for pipeline_ in pipelines:
+            _pull_pipeline_metrics(pipeline_)
+    else:
+
+        jmxes = sdc_client.get_jmxes_async([
+            (pipeline_.streamsets, f'metrics:name=sdc.pipeline.{pipeline_.name}.*',)
+            for pipeline_ in pipelines])
+        for pipeline_, jmx in zip(pipelines, jmxes):
+            _pull_pipeline_metrics(pipeline_, jmx)
```

Comment:
As I right understand, that only jmx changes depending on asynchronous flag. So, how about preparing jmxes first and then running pulling.
Like so:

```python
def _process_pipeline_metrics(pipelines: List[pipeline.Pipeline], asynchronous: bool = False) -> None:
    # prepeare jmxes
    if not asynchronous:
        jmxes = [sdc_client.get_jmx(pipeline_.streamsets, f'metrics:name=sdc.pipeline.{pipeline_.name}.*')
                 for pipeline_ in pipelines]
    else:
        jmxes = sdc_client.get_jmxes_async([(pipeline_.streamsets, f'metrics:name=sdc.pipeline.{pipeline_.name}.*',)
                                            for pipeline_ in pipelines])
    # actual pulling
    for pipeline_, jmx in zip(pipelines, jmxes):
        _pull_pipeline_metrics(pipeline_, jmx)
```

This will make to request for jmxes in on function instead of two different